File: backend/api/Models/StudySession.cs
using api.DTO;

namespace api.Models;

public class StudySession
{
    public int Id { get; set; }
    public int UserId { get; set; }
    public int DeckId { get; set; }
    public Deck Deck { get; set; }
    public int Duration { get; set; }
    public int CardCount { get; set; }
    public float AverageEaseFactor { get; set; }
    public float AverageRepetitions { get; set; }
    public DateTime StudiedAt { get; set; }
    public int CardsRated1 { get; set; }
    public int CardsRated2 { get; set; }
    public int CardsRated3 { get; set; }
    public int CardsRated4 { get; set; }
    public int CardsRated5 { get; set; }

    public StudySession() {}

    public StudySession(StudySessionCreateRequest request)
    {
        UserId = request.UserId;
        DeckId = request.DeckId;
        Duration = request.Duration;
        CardCount = request.CardCount;
        AverageEaseFactor = request.AverageEaseFactor;
        AverageRepetitions = request.AverageRepetitions;
        StudiedAt = request.StudiedAt;
        CardsRated1 = request.CardsRated1;
        CardsRated2 = request.CardsRated2;
        CardsRated3 = request.CardsRated3;
        CardsRated4 = request.CardsRated4;
        CardsRated5 = request.CardsRated5;   
    }

    public StudySession Update(StudySessionUpdateRequest request)
    {
        UserId = request.UserId;
        DeckId = request.DeckId;
        Duration = request.Duration;
        CardCount = request.CardCount;
        AverageEaseFactor = request.AverageEaseFactor;
        AverageRepetitions = request.AverageRepetitions;
        StudiedAt = request.StudiedAt;
        CardsRated1 = request.CardsRated1;
        CardsRated2 = request.CardsRated2;
        CardsRated3 = request.CardsRated3;
        CardsRated4 = request.CardsRated4;
        CardsRated5 = request.CardsRated5;
        return this;
    }
}

File: mobile/lib/models/study_session.dart
class StudySession {
  int userId;
  int deckId;
  int duration;
  int cardCount;
  num averageEaseFactor;
  num averageRepetitions;
  DateTime studiedAt;
  int cardsRated1;
  int cardsRated2;
  int cardsRated3;
  int cardsRated4;
  int cardsRated5;

  StudySession({
    required this.userId,
    required this.deckId,
    required this.duration,
    required this.cardCount,
    required this.averageEaseFactor,
    required this.averageRepetitions,
    required this.studiedAt,
    required this.cardsRated1,
    required this.cardsRated2,
    required this.cardsRated3,
    required this.cardsRated4,
    required this.cardsRated5,
  });

  factory StudySession.fromJson(Map<String, dynamic> json) => StudySession(
        userId: json["userId"],
        deckId: json["deckId"],
        duration: json["duration"],
        cardCount: json["cardCount"],
        averageEaseFactor: json["averageEaseFactor"]?.toDouble(),
        averageRepetitions: json["averageRepetitions"]?.toDouble(),
        studiedAt: DateTime.parse(json["studiedAt"]),
        cardsRated1: json["cardsRated1"],
        cardsRated2: json["cardsRated2"],
        cardsRated3: json["cardsRated3"],
        cardsRated4: json["cardsRated4"],
        cardsRated5: json["cardsRated5"],
      );

  Map<String, dynamic> toJson() => {
        "userId": userId,
        "deckId": deckId,
        "duration": duration,
        "cardCount": cardCount,
        "averageEaseFactor": averageEaseFactor,
        "averageRepetitions": averageRepetitions,
        "studiedAt": studiedAt.toIso8601String(),
        "cardsRated1": cardsRated1,
        "cardsRated2": cardsRated2,
        "cardsRated3": cardsRated3,
        "cardsRated4": cardsRated4,
        "cardsRated5": cardsRated5,
      };
}
